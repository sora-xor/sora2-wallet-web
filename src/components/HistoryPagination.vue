<template>
  <s-pagination class="history-pagination" layout="slot" :current-page.sync="currentPage" :page-size="pageAmount">
    <span class="el-pagination__total">{{ totalText }}</span>
    <s-button
      type="link"
      :tooltip="t('history.firstText')"
      :disabled="disabledFirstPrev"
      @click="handlePaginationClick(PaginationButton.First)"
    >
      <s-icon name="chevrons-left-16" size="14" />
    </s-button>
    <s-button
      type="link"
      :tooltip="t('history.prevText')"
      :disabled="disabledFirstPrev"
      @click="handlePaginationClick(PaginationButton.Prev)"
    >
      <s-icon name="chevron-left-16" size="14" />
    </s-button>
    <s-button
      type="link"
      :tooltip="t('history.nextText')"
      :disabled="disabledNextLast"
      @click="handlePaginationClick(PaginationButton.Next)"
    >
      <s-icon name="chevron-right-16" size="14" />
    </s-button>
    <s-button
      type="link"
      :tooltip="t('history.lastText')"
      :disabled="disabledNextLast"
      @click="handlePaginationClick(PaginationButton.Last)"
    >
      <s-icon name="chevrons-right-16" size="14" />
    </s-button>
  </s-pagination>
</template>

<script lang="ts">
import { Mixins, Component, Prop } from 'vue-property-decorator';

import TranslationMixin from './mixins/TranslationMixin';
import { PaginationButton } from '../consts';

@Component
export default class HistoryPagination extends Mixins(TranslationMixin) {
  @Prop({ default: 1, type: Number }) readonly currentPage!: number;
  @Prop({ default: 10, type: Number }) readonly pageAmount!: number;
  @Prop({ default: 0, type: Number }) readonly total!: number;
  @Prop({ default: false, type: Boolean }) readonly loading?: boolean;
  @Prop({ default: 1, type: Number }) readonly lastPage!: number;

  readonly PaginationButton = PaginationButton;

  get totalText(): string {
    const upperNumber = this.pageAmount * this.currentPage;

    return `${this.t('ofText', {
      first: `${upperNumber - this.pageAmount + 1}-${upperNumber > this.total ? this.total : upperNumber}`,
      second: this.total,
    })}`;
  }

  /** Disable First, Prev buttons if this is the first page or loading state */
  get disabledFirstPrev(): boolean {
    return this.currentPage === 1 || !!this.loading;
  }

  /** Disable Next, Last buttons if this is the last page or loading state */
  get disabledNextLast(): boolean {
    return this.currentPage === this.lastPage || !!this.loading;
  }

  handlePaginationClick(button: PaginationButton): void {
    this.$emit('pagination-click', button);
  }
}
</script>

<style lang="scss">
.history-pagination.el-pagination {
  margin-top: #{$basic-spacing-medium};
  display: flex;
  justify-content: end;
  align-items: baseline;
  padding-left: 0;
  padding-right: 0;
  .el-pagination__total {
    margin-right: auto;
    letter-spacing: var(--s-letter-spacing-small);
    color: var(--s-color-base-content-secondary);
  }
  .el-button.neumorphic {
    margin-left: 0;
    height: var(--s-small-medium);
    padding: 0;
    &:not(:hover):not(:active) {
      color: var(--s-color-base-content-tertiary);
    }
    span {
      min-width: calc(var(--s-basic-spacing) * 2.5);
    }
  }
}
</style>
