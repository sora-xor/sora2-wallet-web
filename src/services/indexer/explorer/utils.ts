import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
