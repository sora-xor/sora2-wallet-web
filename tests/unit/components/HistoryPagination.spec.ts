import omit from 'lodash/fp/omit';

import { useDescribe, useShallowMount } from '../../utils';
import { MOCK_HISTORY_PAGINATION } from '../../utils/HistoryPaginationMock';

import HistoryPagination from '@/components/HistoryPagination.vue';

useDescribe('HistoryPagination.vue', HistoryPagination, () => {
  MOCK_HISTORY_PAGINATION.map((item) => {
    it(`[${item.title}]: should be rendered correctly`, () => {
      const propsData = omit(['title'], item);
      const wrapper = useShallowMount(HistoryPagination, { propsData });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
