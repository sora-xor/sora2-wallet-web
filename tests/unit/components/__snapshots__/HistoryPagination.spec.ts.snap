// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryPagination.vue [With Empty History]: should be rendered correctly 1`] = `
<s-pagination-stub
  borderradius="small"
  class="history-pagination"
  currentpage="1"
  layout="slot"
  nexttext=""
  oftext="of"
  pagesize="10"
  pagesizes="10,20,30,40,50,100"
  perpagetext="Rows per page"
  popperclass=""
  prevtext=""
  totaltext="0-0 of 0"
>
  <span
    class="el-pagination__total"
  >
    0-0 of 0
  </span>
   
  <s-button-stub
    borderradius="small"
    disabled="true"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="First"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevrons-left-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    disabled="true"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Previous"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevron-left-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Next"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevron-right-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Last"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevrons-right-16"
      size="14"
    />
  </s-button-stub>
</s-pagination-stub>
`;

exports[`HistoryPagination.vue [With Last Page]: should be rendered correctly 1`] = `
<s-pagination-stub
  borderradius="small"
  class="history-pagination"
  currentpage="2"
  layout="slot"
  nexttext=""
  oftext="of"
  pagesize="10"
  pagesizes="10,20,30,40,50,100"
  perpagetext="Rows per page"
  popperclass=""
  prevtext=""
  totaltext="11-11 of 11"
>
  <span
    class="el-pagination__total"
  >
    11-11 of 11
  </span>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="First"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevrons-left-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Previous"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevron-left-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    disabled="true"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Next"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevron-right-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    disabled="true"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Last"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevrons-right-16"
      size="14"
    />
  </s-button-stub>
</s-pagination-stub>
`;

exports[`HistoryPagination.vue [With Second Page]: should be rendered correctly 1`] = `
<s-pagination-stub
  borderradius="small"
  class="history-pagination"
  currentpage="2"
  layout="slot"
  nexttext=""
  oftext="of"
  pagesize="10"
  pagesizes="10,20,30,40,50,100"
  perpagetext="Rows per page"
  popperclass=""
  prevtext=""
  totaltext="11-20 of 23"
>
  <span
    class="el-pagination__total"
  >
    11-20 of 23
  </span>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="First"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevrons-left-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Previous"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevron-left-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Next"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevron-right-16"
      size="14"
    />
  </s-button-stub>
   
  <s-button-stub
    borderradius="small"
    icon=""
    iconposition="left"
    nativetype="button"
    size="medium"
    tabindex="0"
    tooltip="Last"
    tooltipplacement="top"
    type="link"
    usedesignsystem="true"
  >
    <s-icon-stub
      name="chevrons-right-16"
      size="14"
    />
  </s-button-stub>
</s-pagination-stub>
`;
